content/blog/client-futures.md

```diff
@@ -0,0 +1,72 @@
+---
+title: Migrating the TiKV Rust client from futures 0.1 to 0.3
+date: 2019-07-09
+author: Nick Cameron
+---
+
+This article was originally posted at [Featherweight Musings](https://www.ncameron.org/blog/migrating-a-crate-from-futures-0-1-to-0-3/).
+
+I recently migrated a small/medium-sized crate from Futures 0.1 to 0.3. It was fairly easy, but there were some tricky bits and some things that were not well documented, so I think it is worth me writing up my experience.
+
+The crate I migrated is the [Rust client](https://github.com/tikv/client-rust) for the TiKV database. It is about 5500 LoC and uses futures fairly heavily (because it communicates with TiKV using gRPC and the interface to that is async).
+
+Async programming in Rust is a large area and has been in development for some time. One of the core parts is the [futures](https://github.com/rust-lang-nursery/futures-rs) library. That is a standard Rust library which provides data types and functionality for programming using futures. It is essential for async programming, but not everything you need - you also need libraries for driving the event loop and interacting with the operating system.
+
+Over the years, the futures library has changed a lot. A lot of older code was developed using the 0.1 series of releases and has not been updated. The more recent versions are the 0.3 series (futures-preview on crates.io). The reason for this divergence is that there are a lot of changes between 0.1 and 0.3, 0.1 is fairly stable, whereas 0.3 has been evolving rapidly. Long-term, 0.3 will turn into 1.0; parts of the library are moving into std, and the first parts have recently been stabilised (e.g., the `Future` trait).
+
+We want the Rust client to work with stable compilers, so we limited the core library to only use features which are stable or would be soon. We did use async/await in our documentation and examples since it is so much more ergonomic, and will eventually be the recommended way to program asynchronously in Rust. As well as avoiding async/await in our library, we also depend on crates which use futures 0.1, which means we needed to use the compatibility layer a lot. Therefore, this might not be a totally typical migration.
```

Comment:
```suggestion
We want the Rust client to work with stable compilers, so we limited the core library to only use features which are stable or which will be soon. We did use async/await in our documentation and examples since it is so much more ergonomic, and will eventually be the recommended way to program asynchronously in Rust. In addition to avoiding async/await in our library, we also depend on crates which use futures 0.1, which means we needed to use the compatibility layer a lot. Therefore, this might not be a totally typical migration.
```

Reply:
1. A little confused about the shifts of verb tenses in the paragraph. Are we talking about the migration process in the past, or something we'd rather refer to in the generic present?
2. On second thought, I think  "which are stable or will be soon" should be fine.